Dropdown/DropdownList.js

```diff
@@ -180,142 +180,144 @@ const ReadyState = {
+				prevChildren: props.children,
 				prevFocused: adjustedFocusIndex,
-				prevSelected: this.props.selected,
-				prevSelectedKey: getKey(this.props),
+				prevSelected: props.selected,
+				prevSelectedKey: getKey(props),
 				ready: ReadyState.INIT
 			});
-		}
+		}, [props]);
 
-		scrollIntoView = () => {
-			let {selected} = this.props;
+		const scrollIntoView = useCallback(() => {
+			let {selected} = props;
 
-			if (this.state.prevFocused == null && !isSelectedValid(this.props)) {
+			if (state.prevFocused == null && !isSelectedValid(props)) {
 				selected = 0;
-			} else if (this.state.prevFocused != null) {
-				selected = this.state.prevFocused;
+			} else if (state.prevFocused != null) {
+				selected = state.prevFocused;
 			}
 
-			this.scrollTo({
+			scrollToRef.current({
 				animate: false,
 				focus: true,
 				index: selected,
 				offset: ri.scale(126 * 2), // @sand-item-small-height * 2 (TODO: large text mode not supported!)
 				stickTo: 'start' // offset from the top of the dropdown
 			});
 
-			this.setState({ready: ReadyState.SCROLLED});
-		};
+			setState(value => {
+				return {...value, ready: ReadyState.SCROLLED};
+			});
+		}, [props, state.prevFocused]);
 
-		focusSelected () {
-			this.setState({ready: ReadyState.DONE});
-		}
+		useEffect(() => {
+			if (state.ready === ReadyState.INIT) {
+				scrollIntoView();
+			} else if (state.ready === ReadyState.SCROLLED) {
+				focusSelected();
+			} else {
+				const key = getKey(props);
+				const keysDiffer = key && state.prevSelectedKey && key !== state.prevSelectedKey;
+
+				if (keysDiffer ||
+					((!key || !state.prevSelectedKey) && state.prevSelected !== props.selected) ||
+					!compareChildren(state.prevChildren, props.children)
+				) {
+					resetFocus(keysDiffer);
+				}
+			}
+		}, [props, resetFocus, scrollIntoView, state]);
 
-		handleFocus = (ev) => {
+		const setScrollTo = useCallback((scrollTo) => {
+			scrollToRef.current = scrollTo;
+		}, []);
+
+		const handleFocus = useCallback((ev) => {
 			const current = ev.target;
-			const dropdownListNode = this.clientSiblingRef?.current;
+			const dropdownListNode = clientSiblingRef?.current;
 
-			if (this.state.ready === ReadyState.DONE && !Spotlight.getPointerMode() &&
+			if (state.ready === ReadyState.DONE && !Spotlight.getPointerMode() &&
 				current.dataset['index'] != null && dropdownListNode.contains(current)
 			) {
 				const focusedIndex = Number(current.dataset['index']);
-				const lastFocusedKey = getKey({children: this.props.children, selected: focusedIndex});
-				this.lastFocusedKey = lastFocusedKey;
+				lastFocusedKey.current = getKey({children: props.children, selected: focusedIndex});
 			}
 
-			if (this.props.onFocus) {
-				this.props.onFocus(ev);
+			if (props.onFocus) {
+				props.onFocus(ev);
 			}
-		};
+		}, [props, state.ready]);
 
-		render () {
-			const props = {...this.props};
-			delete props.handleSpotlightPause;
+		const newProps = Object.assign({}, props);
```

Comment:
let's rename this to wrappedComponentProps

Reply:
Done